docs/rag/ilab-rag-retrieval.md

```diff
@@ -0,0 +1,455 @@
+ in Haystack.
+* [MilvusEmbeddingRetriever](https://github.com/milvus-io/milvus-haystack/blob/77b27de00c2f0278e28b434f4883853a959f5466/src/milvus_haystack/milvus_embedding_retriever.py#L18)
+
+### 2.10 Workflow Visualization
+
+Embedding ingestion pipeline:
+![ingestion-mvp](./images/ingestion-mvp.png)
+RAG-based Chat pipeline:
+![rag-chat](./images/rag-chat.png)
+
+### 2.11 Proposed Implementation Stack
+
+> **ℹ️ Note:** This stack is still under review. The proposed list represents potential candidates based on the current state of discussions.
+
+The following technologies form the foundation of the proposed solution:
+
+* [Haystack](https://haystack.deepset.ai/): Framework for implementing RAG pipelines and applications.
+* [MilvusLite](https://milvus.io/docs/milvus_lite.md): The default vector database for efficient storage and retrieval of embeddings.
+* [Docling](https://github.com/DS4SD/docling): Document processing tool. For more details, refer to William’s blog, [Docling: The missing document processing companion for generative AI](https://www.redhat.com/en/blog/docling-missing-document-processing-companion-generative-ai).
+
+## 3. Design Considerations
+
+* As decided in [PR #165](https://github.com/instructlab/dev-docs/pull/165), functions related to RAG ingestion
+  and retrieval are located in the dedicated folder `src/instructlab/rag`.
+* The solution must minimize changes to existing modules by importing the required functions from the
+  `instructlab.rag` package.
+* The solution must adopt a pluggable design to facilitate seamless integration of additional components:
+  * **Vector stores**: Support all selected implementations (e.g., Milvus).
+  * **Embedding models**: Handle embedding models using the appropriate embedder implementation for the
+    chosen framework (e.g., Haystack).
+  * Consider using factory functions to abstract implementations and enhance code flexibility.
+* Optional dependencies for 3rd party integrations should be defined in `pyproject.toml` and documented for
+  clarity. Users can install optional components with commands like:
+
+  `pip install instructlab[milvus]`
+
+  3rd party dependencies may also be grouped in files such as `requirements/milvus.txt`.
+
+## 4. Future Enhancements
+
+### 4.1 Model Evaluation
+
+**TODO** A separate ADR will be defined.
+
+### 4.2 Advanced RAG retrieval steps
+
+* [Ranking retriever's result][ranking]:
+
+```bash
+ilab model chat --rag --ranking --ranking-top-k=5 --ranking-model=cross-encoder/ms-marco-MiniLM-L-12-v2
+```
+
+* [Query expansion][expansion]:
+
+```bash
+ilab model chat --rag --query-expansion --query-expansion-prompt="$QUERY_EXPANSION_PROMPT" --query-expansion-num-of-queries=5
+```
+
+* Using retrieval strategy:
+
+```bash
+ilab model chat --rag --retrieval-strategy query-expansion --retrieval-strategy-options="prompt=$QUERY_EXPANSION_PROMPT;num_of_queries=5"
+```
```

Comment:
This kind of option can make usage more complicated, especially when multiple parameters need to be configured together?

Reply:
I agree.  I think more consideration is needed for how we will enable retrieval strategy configuration.  I think that should be addressed in a future document though; for this document, I would recommend just adding a note indicating that the entire "future enhancements" section is just some preliminary thoughts on the sorts of things we might add in the future.

Reply:
> I would recommend just adding a note indicating that the entire "future enhancements" section is just some preliminary thoughts on the sorts of things we might add in the future.

Agree